keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2165 @@
+
+PodStatus and ContainerStatus types are extended to include information of
+QoS-class resource assignments. The main motivation for this change is to
+communicate to the user "implicit defaults", i.e. report back what was assigned
+for QoS-class resources for which nothing was requested (see
+[implicit defaults](#implicit-defaults))
+
+```diff
+type PodStatus struct {
+@@ -3534,6 +3537,10 @@ type PodStatus struct {
+        // Status for any ephemeral containers that have run in this pod.
+        // +optional
+        EphemeralContainerStatuses []ContainerStatus
++
++       // QOSResources shows the assignment of pod-level QoS resources.
++       // +optional
++       QOSResources []PodQOSResourceRequest
+ }
+```
+
+```diff
+type ContainerStatus struct {
+@@ -2437,6 +2437,9 @@ type ContainerStatus struct {
+        Started     *bool
++
++       // QOSResources shows the assignment of QoS resources.
++       // +optional
++       QOSResources []QOSResourceRequest
+ }
+```
+
+Status will re-use the PodQOSResourceRequest and QOSResourceRequest that are
+specified as part of [PodSpec](#podspec) update.
+
+#### NodeStatus
+
+We extend NodeStatus to list available QoS-class resources on a node, This
+consists of the list of the available QoS-class resource types and the classes
+available within each of these resource types.
+
+```diff
+ type NodeStatus struct {
+@@ -4444,6 +4482,11 @@ type NodeStatus struct {
+        // Status of the config assigned to the node via the dynamic Kubelet config feature.
+        Config *NodeConfigStatus
++       // QOSResources contains information about the QoS resources that are
++       // available on the node.
++       // +featureGate=QOSResources
++       // +optional
++       QOSResources QOSResourceStatus
+ }
+
+...
+
++// QOSResourceStatus describes QoS resources available on the node.
++type QOSResourceStatus struct {
++       // PodQOSResources contains the QoS resources that are available for pods
++       // to be assigned to.
++       PodQOSResources []QOSResourceInfo
++       // ContainerQOSResources contains the QoS resources that are available for
++       // containers to be assigned to.
++       ContainerQOSResources []QOSResourceInfo
++}
+
++// QOSResourceInfo contains information about one QoS resource type.
++type QOSResourceInfo struct {
++       // Name of the resource.
++       Name QOSResourceName
++       // Mutable is set to true if the resource supports in-place updates.
++       Mutable bool
++       // Classes available for assignment.
++       Classes []QOSResourceClassInfo
++}
+
++// QOSResourceClassInfo contains information about single class of one QoS
++// resource.
++type QOSResourceClassInfo struct {
++       // Name of the class.
++       Name string
++       // Capacity is the number of maximum allowed simultaneous assignments into this class
++       // Zero means "infinite" capacity i.e. the usage is not restricted
++       // +optional
++       Capacity int64
++}
+```
+
```

Comment:
Would be nice to see an example Node status serialized out as YAML, including the new fields.